### PR TITLE
1488 create mpConfig-1.2

### LINKED ISSUE #1488: Create mpConfig-1.2 feature structure
Create mpConfig-1.2 feature

### PR DESCRIPTION
**Pull request by tevans78**
#build 
fixes #1488 

Included in this PR is support for the SPI change (issue #1493) which allows Converters to be more precisely registered, with type and priority. Further testing is still required there, particularly around specifying unusual types and using lambdas to implement converters.

Addtional project structure changes may be required as we implement the other config 1.2 enhancements.

### CODE REVIEW

File: dev/com.ibm.websphere.appserver.mpConfig-1.2/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/bin/
```
Comment (aguibert) [inline dev/com.ibm.websphere.appserver.mpConfig-1.2/.gitignore]: file can be removed

File: dev/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/bin/
```
Comment (aguibert) [inline dev/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2/.gitignore]: file can be removed

File: dev/com.ibm.websphere.appserver.org.eclipse.microprofile.config-1.2/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/bin/
```
Comment (aguibert) [inline dev/com.ibm.websphere.appserver.org.eclipse.microprofile.config-1.2/.gitignore]: file can be removed

File: dev/com.ibm.websphere.org.eclipse.microprofile.config.1.2/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/bin/
```
Comment (aguibert) [inline dev/com.ibm.websphere.org.eclipse.microprofile.config.1.2/.gitignore]: file can be removed

File: dev/com.ibm.ws.microprofile.config.1.2_fat/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#*******************************************************************************
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (aguibert) [inline dev/com.ibm.ws.microprofile.config.1.2_fat/bnd.bnd]: copyright headers should be updated to 2018

File: dev/com.ibm.ws.microprofile.config.1.2_fat/bnd.bnd [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+javac.target: 1.8
+	
+fat.project: true
+
+-buildpath: \
+	com.ibm.websphere.javaee.ejb.3.2;version=latest,\
+	com.ibm.websphere.javaee.servlet.4.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.config.1.2;version=latest,\
+	com.ibm.websphere.javaee.annotation.1.3;version=latest,\
+	com.ibm.ws.security;version=latest,\
```
Comment (aguibert) [inline dev/com.ibm.ws.microprofile.config.1.2_fat/bnd.bnd]: these deps can probably be removed from the buildpath and will make the FAT build and run faster locally:
```com.ibm.ws.security;version=latest,\
com.ibm.websphere.security;version=latest,\```

File: dev/com.ibm.ws.microprofile.config.1.2_fat/publish/servers/.gitignore [hunk 7]
```diff
@@ -0,0 +1,5 @@
+*/dropins/*
```
Comment (aguibert) [inline dev/com.ibm.ws.microprofile.config.1.2_fat/publish/servers/.gitignore]: this file can be removed

File: dev/com.ibm.ws.microprofile.config.1.2_fat/test-applications/converterApp/src/com/ibm/ws/microprofile/config12/converter/priority/web/ConverterPriorityServlet.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+@SuppressWarnings("serial")
+@WebServlet("/ConverterPriorityServlet")
+public class ConverterPriorityServlet extends FATServlet {
+
+    @Inject
+    ConverterPriorityBean bean;
+
+    @Test
+    @Mode(TestMode.LITE)
```
Comment (aguibert) [inline dev/com.ibm.ws.microprofile.config.1.2_fat/test-applications/converterApp/src/com/ibm/ws/microprofile/config12/converter/priority/web/ConverterPriorityServlet.java]: don't need to annotate for `@Mode(TestMode.LITE)`, that is the default value if no `@Mode` annotation is present

File: dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java [hunk 9]
```diff
@@ -57,13 +56,68 @@ public static FeatureReplacementAction EE8_FEATURES() {
...
     }
 
     public FeatureReplacementAction(Set<String> removeFeatures, Set<String> addFeatures) {
-        this.removeFeatures.addAll(removeFeatures);
+        this();
+        addFeatures(addFeatures);
+        removeFeatures(removeFeatures);
+    }
+
+    public FeatureReplacementAction() {
```
Comment (aguibert) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java]: If we are going to add this ctor that's not valid unless you call other methods, lets add a check to `setup()` that validates that some features additions/removals have been configured (i.e. blow up if `addFeatures.size() == 0 && removeFeatures.size() == 0`)

**Comment by tevans78** [comment]
#build

**Comment by tevans78** [comment]
#build

**Comment by tevans78** [comment]
PB tevans78-1532-20180110-1745 is good, 1 unrelated failure.
https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_eGTYoPYkEeeSXY-prsQAeQ

File: dev/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2.feature [hunk 10]
```diff
@@ -0,0 +1,11 @@
+-include= ~../cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.mpConfig1.2-cdi1.2
```
Comment (Emily-Jiang) [inline dev/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2.feature]: Can you remove the cdi version from the symbolic name?

Comment (tevans78) [reply dev/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2/com.ibm.websphere.appserver.mpConfig1.2-cdi1.2.feature]: OK, will do in a follow up PR

File: dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java (hunk 9 above)
Comment (Emily-Jiang) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java]: I don't quite understand the purpose of this ctor.

**Comment by tevans78** [comment]
@Emily-Jiang I'm not sure that '+' is a valid character all the way through the stack?

**Pull request merged**